pkg/network/tracer/tracer_linux_test.go

```diff
@@ -2068,3 +2069,67 @@ func TestBlockingReadCounts(t *testing.T) {
 
 	assert.Equal(t, uint64(n), conn.MonotonicSum().RecvBytes)
 }
+
+// GoTLS test
+
+func TestHTTPGoTLSCaptureNewProcess(t *testing.T) {
+	const ClientBin string = "./testutil/cmd/gotls_client/gotls_client"
+	const ServerAddr string = "localhost:8081"
+	const ExpectedOccurrences int = 10
+
+	if !httpSupported(t) {
+		t.Skip("HTTPS feature not available on pre 4.14.0 kernels")
+	}
+
+	if !httpsSupported(t) {
+		t.Skip("HTTPS feature not available supported for this setup")
+	}
+
+	// Given
+	var closeServer func() = httptest.HTTPServer(t, ServerAddr, httptest.Options{
+		EnableTLS: true,
+	})
+
+	cfg := config.New()
+	cfg.EnableHTTPMonitoring = true
+	cfg.EnableHTTPSMonitoring = true
+	cfg.BPFDebug = true
+
+	tr, err := NewTracer(cfg)
+	require.NoError(t, err)
+	defer tr.Stop()
+	require.NoError(t, tr.RegisterClient("1"))
+
+	// When
+	req, err := nethttp.NewRequest(nethttp.MethodGet, fmt.Sprintf("https://%s/%d/request", ServerAddr, nethttp.StatusOK), nil)
+	require.NoError(t, err)
+
+	clientBuildCmd := fmt.Sprintf("go build -o %s ./testutil/cmd/gotls_client", ClientBin)
+	clientCmd := fmt.Sprintf("%s %s %d", ClientBin, ServerAddr, ExpectedOccurrences)
+	t.Cleanup(func() { os.RemoveAll(ClientBin) })
+	_ = testutil.RunCommands(t, []string{clientBuildCmd, clientCmd}, false)
+
+	closeServer()
+
+	// Then
+	occurrences := 0
+	require.Eventually(t, func() bool {
+		stats, err := tr.GetActiveConnections("1")
+		require.NoError(t, err)
+		occurrences += countRequestOccurrences(t, stats, req)
+		return occurrences == ExpectedOccurrences
+	}, 3*time.Second, 100*time.Millisecond, "Expected to find the request %v times, got %v captured", ExpectedOccurrences, occurrences)
```

Comment:
`require.Eventually` calculates the value of `"Expected to find the request %v times, got %v captured", ExpectedOccurrences, occurrences` before it starts to run, so `got %v captured` will always show `got 0 captured` although you might have captured 9 requests

Try running the following and see the error being reported
```
func TestBla(t *testing.T) {
	ExpectedOccurrences := 10
	occurrences := 0
	require.Eventually(t, func() bool {
		if occurrences == 0 {
			occurrences = 1
		}

		return occurrences == ExpectedOccurrences
	}, 3*time.Second, 100*time.Millisecond, "Expected to find the request %v times, got %v captured", ExpectedOccurrences, occurrences)
}
```

Reply:
So I dived into it a little, and figured out what happens here. Things gets confusing because of the closure that modifies `occurences`, so it makes us think `Eventually` computes the string before it starts running.
However, when looking at the implementation of `Eventually` it does compute the string only on failure. The real culprit here is that Go passes arguments by value, so eventually receives *the value* of `occurences`, which can only be `0`.
In an ideal world, we would want to pass a reference to `occurences` (kinda like you would in C++ or Rust) so that when it formats the string, it dereferences it and get the actual value of `occurences`. Sadly, Go is not an ideal world, so all we get are pointer types. Trying to pass `&occurences` to `Eventually` will not work, because then it will format the pointer's value, instead of the value pointed to by the pointer.
Go also does not allow us to write the `func (i *int) String() string` function because `int` is a base type.

We can work around that, with a new type wrapping `int` for which we implement the `Stringer` interface. If I rewrite your example with this approach we get this:

```go
type PrintableInt int

func (i *PrintableInt) String() string {
	if i == nil {
		return "nil"
	}

	return fmt.Sprintf("%d", *i)
}

func TestBla(t *testing.T) {
	ExpectedOccurrences := PrintableInt(10)
	occurrences := PrintableInt(0)
	require.Eventuallyf(t, func() bool {
		if occurrences == 0 {
			occurrences = 1
		}

		return occurrences == ExpectedOccurrences
	}, 3*time.Second, 100*time.Millisecond, "Expected to find the request %v times, got %v captured", ExpectedOccurrences, &occurrences)
}
```
Which then gives us the expected output:
```
Messages:       Expected to find the request 10 times, got 1 captured
```

Would this be an acceptable solution?

Reply:
Your solution works, please change the code